src/OpenApi/src/Services/OpenApiDocumentService.cs

```diff
@@ -296,11 +296,78 @@ private OpenApiRequestBody GetFormRequestBody(IList<ApiRequestFormat> supportedR
             Content = new Dictionary<string, OpenApiMediaType>()
         };
 
-        // Forms are represented as objects with properties for each form field.
         var schema = new OpenApiSchema { Type = "object", Properties = new Dictionary<string, OpenApiSchema>() };
-        foreach (var parameter in formParameters)
+        // Group form parameters by their parameter name because MVC explodes form parameters that are bound from the
+        // same model instance into separate parameters, while minimal APIs does not.
+        //
+        // public record Todo(int Id, string Title, bool Completed, DateTime CreatedAt)
+        // public void PostMvc([FromForm] Todo person) { }
+        // app.MapGet("/form-todo", ([FromForm] Todo todo) => Results.Ok(todo));
+        //
+        // In the example above, MVC will bind four separate arguments to the Todo model while minimal APIs will
+        // bind a single Todo model instance to the todo parameter. Grouping by name allows us to handle both cases.
+        var groupedFormParameters = formParameters.GroupBy(parameter => parameter.ParameterDescriptor.Name).ToList();
```

Comment:
Is the `ToList` just to make pattern matching work?

Reply:
I did it because I was enumerating the collection a couple of times. Although that's negligible since I've reacted to the feedback [here](https://github.com/dotnet/aspnetcore/pull/55321#discussion_r1576996250).